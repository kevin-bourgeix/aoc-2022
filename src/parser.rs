use std::fs;

use regex::Regex;

// Parses an aoc file into a vector of strings
// By defaults it splits on newlines, but it can be changed easily
pub fn parse_aoc_file(filename: &str, delimiter: Option<&str>) -> Vec<String> {
    let Ok(file) = fs::read_to_string(filename) else {
        panic!("Could not read file {filename}");
    };
    let split: Vec<&str> = file.split(delimiter.unwrap_or("\n")).collect();

    return split.iter().map(|s| s.to_string()).collect();
}

// Parses strings with the given parser function
pub fn parse_into_struct<T>(data: Vec<String>, parser: fn(String) -> T) -> Vec<T> {
    let mut result = Vec::new();
    for line in data {
        result.push(parser(line));
    }
    result
}

// Reads through a vector of lines while not matching a specific regex
pub fn parse_until_pattern(data: Vec<String>, pattern: &str) -> (Vec<String>, Vec<String>) {
    let Ok(re) = Regex::new(pattern) else {
        panic!("Could not parse pattern {pattern} at parse_until_pattern");
    };
    let mut index = 0;
    for line in &data {
        if re.is_match(line) {
            break;
        }
        index += 1;
    }
    (data[0..index].to_vec(), data[index..].to_vec())
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;

    #[derive(Debug)]
    struct TestStruct {
        a: String,
        b: String,
    }

    #[test]
    fn test_parse_aoc_file_default() {
        let data = parse_aoc_file("./src/tests/numbers1.txt", None);
        insta::assert_debug_snapshot!(data);
    }

    #[test]
    fn test_parse_aoc_file_comma() {
        let data = parse_aoc_file("./src/tests/numbers2.txt", Some(","));
        insta::assert_debug_snapshot!(data);
    }

    #[test]
    fn test_parse_aoc_file_with_blank() {
        let data = parse_aoc_file("./src/tests/numbers3.txt", None);
        insta::assert_debug_snapshot!(data);
    }

    #[test]
    fn test_parse_into_struct() {
        let data = vec!["1,2".to_string(), "3,4".to_string()];
        let result = parse_into_struct(data, |s| {
            let split: Vec<&str> = s.split(',').collect();
            TestStruct {
                a: split[0].to_string(),
                b: split[1].to_string(),
            }
        });

        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_into_simple() {
        let data = vec!["1".to_string(), "2".to_string()];
        let result = parse_into_struct(data, |s| s.parse::<i32>().unwrap());

        insta::assert_debug_snapshot!(result);
    }

    #[test]
    fn test_parse_until() {
        let data = vec!["aaaa", "aaaa", "aaaa", "bbbb", "bbbb", "cccc"]
            .iter()
            .map(|s| s.to_string())
            .collect();
        let (first, second) = parse_until_pattern(data, "bbbb");

        insta::assert_debug_snapshot!(vec![first, second]);
    }

    #[test]
    fn test_parse_empty() {
        let data = vec!["1", "", "2", "3"]
            .iter()
            .map(|s| s.to_string())
            .collect();
        let (first, second) = parse_until_pattern(data, r"^$");

        insta::assert_debug_snapshot!(vec![first, second]);
    }
}
