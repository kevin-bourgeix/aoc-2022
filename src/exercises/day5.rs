use regex::Regex;

use crate::parser::{parse_aoc_file, parse_until_pattern};

#[derive(Debug)]
struct Instruction {
    number: usize,
    from: usize,
    to: usize,
}

impl Instruction {
    fn from(instruction: String) -> Instruction {
        let Ok(re) = Regex::new(r"move (\d*) from (\d*) to (\d*)") else {
            panic!("Could not parse regex pattern at Instruction::from");
        };
        let matches = re
            .captures(&instruction)
            .unwrap_or_else(|| panic!("Could not parse instruction {instruction}"));

        Instruction {
            number: matches
                .get(1)
                .unwrap_or_else(|| panic!("Unable to read number from instruction {instruction}"))
                .as_str()
                .parse()
                .unwrap_or_else(|e| {
                    panic!("Unable to parse number from instruction {instruction} with error : {e}")
                }),
            from: matches
                .get(2)
                .unwrap_or_else(|| panic!("Unable to read 'from' from instruction {instruction}"))
                .as_str()
                .parse::<usize>()
                .unwrap_or_else(|e| {
                    panic!("Unable to parse 'from' from instruction {instruction} with error : {e}")
                })
                - 1,
            to: matches
                .get(3)
                .unwrap_or_else(|| panic!("Unable to read 'to' from instruction {instruction}"))
                .as_str()
                .parse::<usize>()
                .unwrap_or_else(|e| {
                    panic!("Unable to parse 'to' from instruction {instruction} with error : {e}")
                })
                - 1,
        }
    }

    fn execute(&self, stacks: &mut [Vec<char>]) {
        let mut i = 0;
        loop {
            let from = stacks
                .get_mut(self.from)
                .unwrap_or_else(|| panic!("Unable to get stack {}", self.from));
            if i == self.number || from.is_empty() {
                break;
            }

            if let Some(item) = from.pop() {
                stacks
                    .get_mut(self.to)
                    .unwrap_or_else(|| panic!("Unable to get stack {}", self.to))
                    .push(item);
            }

            i += 1;
        }
    }

    fn execute_same_order(&self, stacks: &mut [Vec<char>]) {
        // Moving one stack is the same as above function
        if self.number <= 1 {
            self.execute(stacks);
            return;
        }

        let from = stacks
            .get_mut(self.from)
            .unwrap_or_else(|| panic!("Unable to get stack {}", self.from));
        let temp: &mut Vec<char> = &mut vec![];

        for _ in 0..self.number {
            if let Some(item) = from.pop() {
                temp.push(item);
            }
        }

        for _ in 0..self.number {
            if let Some(item) = temp.pop() {
                stacks
                    .get_mut(self.to)
                    .unwrap_or_else(|| panic!("Unable to get stack {}", self.to))
                    .push(item);
            }
        }
    }
}

fn parse_stacks(filename: &str) -> (Vec<Vec<char>>, Vec<String>) {
    let lines = parse_aoc_file(filename, None);
    let (mut stack_lines, remaining) = parse_until_pattern(lines, "^$");
    stack_lines.reverse();

    // Get first line
    let (first_line, stack_lines) = stack_lines
        .split_first()
        .unwrap_or_else(|| panic!("Unable to get first line from {filename}"));

    // Get the total number of stacks to compute
    let num_stacks = first_line
        .split_whitespace()
        .map(|s| {
            s.parse::<u32>()
                .unwrap_or_else(|_| panic!("Unable to parse {s} on the first line (not a number)"))
        })
        .max()
        .unwrap_or(0);

    // Init the stacks
    let mut stacks = Vec::<Vec<char>>::new();
    for _ in 0..num_stacks {
        stacks.push(Vec::new());
    }

    // Fill the stacks
    for line in stack_lines {
        let mut chars = line.chars();
        let i = 1;
        for j in 0..num_stacks {
            let char_index = (i + 4u32 * j) as usize;
            let char = chars.nth(if j == 0 { 1 } else { 3 }).unwrap_or_else(|| {
                panic!("Unable to get char at index {char_index} on line {line}")
            });
            if char != ' ' {
                stacks
                    .get_mut(j as usize)
                    .unwrap_or_else(|| panic!("Stack {j} does not exists"))
                    .push(char);
            }
        }
    }

    (stacks, remaining[1..].to_vec())
}

pub fn day_5_1(filename: &str) -> Vec<char> {
    let (mut stacks, instructions) = parse_stacks(filename);

    for instruction in instructions {
        Instruction::from(instruction).execute(&mut stacks);
    }

    stacks
        .iter()
        .map(|s| s.last().unwrap_or(&' '))
        .filter(|c| **c != ' ')
        .cloned()
        .collect()
}

pub fn day_5_2(filename: &str) -> Vec<char> {
    let (mut stacks, instructions) = parse_stacks(filename);

    for instruction in instructions {
        Instruction::from(instruction).execute_same_order(&mut stacks);
    }

    stacks
        .iter()
        .map(|s| s.last().unwrap_or(&' '))
        .filter(|c| **c != ' ')
        .cloned()
        .collect()
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::{parse_stacks, Instruction};

    #[test]
    fn test_parse_stacks() {
        let parsed = parse_stacks("src/files/day5_1.test");
        assert_debug_snapshot!(parsed);
    }

    #[test]
    fn test_instruction_parsing() {
        let raw = "move 1 from 1 to 2".to_string();
        let instruction = super::Instruction::from(raw);
        assert_debug_snapshot!(instruction);
    }

    #[test]
    fn test_instruction_execute() {
        let mut stacks = vec![vec!['a'], vec![]];
        let instruction = Instruction::from("move 1 from 1 to 2".to_string());
        instruction.execute(&mut stacks);

        assert_debug_snapshot!(stacks);
    }

    #[test]
    fn test_day_5_1() {
        let result = super::day_5_1("src/files/day5_1.test");
        assert_debug_snapshot!(result);
    }

    #[test]
    fn test_day_5_2() {
        let result = super::day_5_2("src/files/day5_1.test");
        assert_debug_snapshot!(result);
    }
}
