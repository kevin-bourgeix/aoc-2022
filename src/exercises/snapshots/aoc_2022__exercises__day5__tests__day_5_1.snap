---
source: src/exercises/day5.rs
assertion_line: 155
expression: result
---
[
    'C',
    'M',
    'Z',
]
